### PR TITLE
Improvements to the windowers

### PR DESCRIPTION
**Pull request by hubertjb**
This PR introduces some minor improvements to the windowers:

- fixing windower functions when `first_samp` is not 0 (i.e., the Raw object has been cropped)
- adding `epochs_kwargs` to windowing functions
- modifying default value of `stop_offset_samples` to None for consistency
- improving docstrings
- adding random state argument to `create_mne_dummy_raw`
- adding tests

### CODE REVIEW

File: braindecode/datautil/windowers.py
Comment (agramfort) [inline braindecode/datautil/windowers.py:171]: i don't like kwargs. I would expose and document here the Epochs parameters you think can be useful.

File: braindecode/datautil/xy.py
Comment (agramfort) [inline braindecode/datautil/xy.py:72]: API change?

File: braindecode/util.py
Comment (agramfort) [inline braindecode/util.py:256]: use sklearn.utils.check_random_state

File: test/unit_tests/datautil/test_windowers.py
Comment (agramfort) [inline test/unit_tests/datautil/test_windowers.py:19]: from xxx.preprocess import preprocess 

I don't like that as it's then confusing what is a function and what is a module

File: braindecode/datautil/windowers.py
Comment (hubertjb) [reply braindecode/datautil/windowers.py]: Fixed in new commit.

File: braindecode/datautil/xy.py
Comment (hubertjb) [reply braindecode/datautil/xy.py:72]: That's correct. Is there a recommended way to flag API changes? I added a warning here: https://github.com/braindecode/braindecode/pull/152/files#diff-c031004527cfc6287fbb1741b7f3e1ddR231

File: test/unit_tests/datautil/test_windowers.py
Comment (hubertjb) [reply test/unit_tests/datautil/test_windowers.py:19]: What about we rename `braindecode.datautil.preprocess` to `braindecode.datautil.preprocessing`? It's a bit longer, but at least we avoid this ambiguity. @robintibor @gemeinl

**Comment by agramfort** [comment]
>
> What about we rename braindecode.datautil.preprocess to
> braindecode.datautil.preprocessing? It's a bit longer, but at least we
> avoid this ambiguity. @robintibor <https://github.com/robintibor> @gemeinl
> <https://github.com/gemeinl>
>

3 levels of modules seem too much for a small package as it is now. I would
expose it
in  braindecode.datautil or consider a new
submodule braindecode.preprocessing

File: test/unit_tests/datautil/test_windowers.py
Comment (gemeinl) [reply test/unit_tests/datautil/test_windowers.py:19]: Personally, I don't like the name "datautil" at all and feel like we could use some major refactoring here. My thoughts:
Currently, ``braindecode.datautil`` for me contains a mess of 
- Loading and saving datasets (``serialization.py``) 
- Creating datasets (``xy.py``, ``mne.py``)
- Preprocessing (``preprocess.py``)
- Creating windows ((``windowers.py``) which is also a bad name, since we don't have windower classes anymore). Also, creating windows is a kind of preprocessing

We could probably rename ``braindecode.datautil`` to ``braindecode.preprocessing`` with ``preprocessing.py`` which contains a merge of ``preprocess.py`` and ``windowers.py``.

We could add ``braindecode.serialization`` with ``serialization.py``

I am not so sure what to do about ``xy.py`` and ``mne.py``, though.

File: braindecode/datautil/windowers.py
Comment (robintibor) [inline braindecode/datautil/windowers.py:117]: This seems potentially quite errorprone code, do we have a test for this?

File: braindecode/datautil/windowers.py
Comment (robintibor) [inline braindecode/datautil/windowers.py:128]: here does not seem to account for trial stop samples?

File: braindecode/datautil/windowers.py
Comment (robintibor) [inline braindecode/datautil/windowers.py:36]: is doc here maybe wrong/inconsistent? 
Below it says:
` trial_stop_offset_samples -       Stop offset from original trial stop, in samples.`
This is also what happens in code in _compute_window_inds:
https://github.com/braindecode/braindecode/blob/fac6cea5d6ea0af49dcb39c85d3be33dfa791c22/braindecode/datautil/windowers.py#L290-L294

File: braindecode/datautil/windowers.py
Comment (robintibor) [inline braindecode/datautil/windowers.py:241]: yeah great, not even sure what was logic behind previous decision. seems quite strange. I wonder actually if we had intended stop offset samples to refer to offset to end of recording? as it is in from events thingy and end of trial?

**Comment by robintibor** [comment]
Did you see the comments @hubertjb ?

**Comment by hubertjb** [comment]
> Did you see the comments @hubertjb ?

@robintibor yes, sorry for the delay! I'm getting back to it today or tomorrow.

File: braindecode/datautil/windowers.py
Comment (hubertjb) [reply braindecode/datautil/windowers.py:117]: Yes, `test_windowers.py::test_windows_cropped` used to test this indirectly - I'll add something more specific to this catch.

Comment (hubertjb) [reply braindecode/datautil/windowers.py:117]: Actually I'm not 100% sure the behaviour of `mne.Epochs` is correct when the raw file has been cropped. Maybe there's something I'm missing that would make the handling of `raw.first_samp` simpler... I opened an issue here https://github.com/mne-tools/mne-python/issues/8201 just in case.

File: braindecode/datautil/windowers.py
Comment (hubertjb) [reply braindecode/datautil/windowers.py:36]: Good point, I'll fix the docstring.

File: braindecode/datautil/windowers.py
Comment (hubertjb) [reply braindecode/datautil/windowers.py:241]: Aah you have a good point, that's probably what was going on. I think the new approach might be more useful: let's say you want to extract windows from 10 minutes of a recording - you can provide the onset and offset as 2 and 12 mins, which is more natural than 2 and *calculate length of recording* - (2 + 10). What do you think?

Comment (robintibor) [reply braindecode/datautil/windowers.py:252]: I agree that extracting a particular length if you don't have events might be what you want in many cases and would also keep like this for this PR. Only question long-term is if we want to have both options (stop given from beginning of recording and from end of recording), one could also do it in every dirty way: if negative, assume it is from end of recording, if positive, assume from start of recording... but as said let's keep in this way (from start of recording) for this PR

File: braindecode/datautil/windowers.py
Comment (robintibor) [reply braindecode/datautil/windowers.py]: So now this will be correct, but need newest mne from master? just for clarification

**Review by robintibor — APPROVED** [review]

**Comment by robintibor** [comment]
so seems ready to merge from what I see @hubertjb ?

**Comment by hubertjb** [comment]
> so seems ready to merge from what I see @hubertjb ?

Almost! I just need to test with the latest MNE as you said - we might indeed need to ask for the latest version until the next version is out.

**Comment by hubertjb** [comment]
@robintibor the latest commit takes the fix in MNE into account.

I guess we should update the requirements to ask for the latest MNE master (or should we wait for an MNE release)? @agramfort

**Comment by hubertjb** [comment]
(The travis fail is because of the MNE version.)

**Comment by robintibor** [comment]
When will this be in pip-installable mne @agramfort ?

**Comment by agramfort** [comment]
@robintibor @hubertjb you can change the CI here to use:

pip install -U https://github.com/mne-tools/mne-python/archive/master.zip

new MNE version should be released on sept 15

**Comment by robintibor** [comment]
ah great then we can also make new braindecode version on pip quickly after. And for now just put as alex said, if you don't know how to do it, i can do it for you as well, just let me know @hubertjb

**Comment by hubertjb** [comment]
Tests are passing with the latest MNE master on Travis! Should we wait for the upcoming MNE release before merging?

**Review by agramfort — APPROVED** [review]

**Comment by robintibor** [comment]
well how is ETA now for MNE release? last was "new MNE version should be released on sept 15" @agramfort ? should be soon then I guess?

**Comment by agramfort** [comment]
it's imminent. Next monday at the latest

**Comment by robintibor** [comment]
so do you want to try go back to MNE from pip since new version is there? @hubertjb then I can merge

**Comment by hubertjb** [comment]
> so do you want to try go back to MNE from pip since new version is there? @hubertjb then I can merge

That'd be great @robintibor, thanks! I just reverted the change to the Travis config file, if tests pass we should be good to go.

**Comment by robintibor** [comment]
hm so seems travis fails with old error, any reason for that @hubertjb @agramfort ? Also CircleCI seems quite a bit slower than all previous builds :/ Don't know why that might be

**Comment by hubertjb** [comment]
Weird... could it be a cache thing? https://docs.travis-ci.com/user/caching/#configuration

**Comment by robintibor** [comment]
It seems reason is mne is still at 0.20.8 at https://anaconda.org/conda-forge/mne . When will this change @agramfort ?

**Comment by agramfort** [comment]
I don't maintain mne in conda. Please use pip

>

**Comment by robintibor** [comment]
ok can you move mne to pip in https://github.com/braindecode/braindecode/blob/master/environment.yml ? I guess that should work then?

**Comment by robintibor** [comment]
great  work @hubertjb ! merged!

**Pull request merged**
